sdk/python/kfp/_local_client.py

```diff
@@ -0,0 +1,441 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import datetime
+import json
+import logging
+import re
+import subprocess
+from typing import Any, Callable, Dict, List, Mapping, Tuple, Union, cast
+
+import kfp
+
+
+class _Dag:
+    def __init__(self, nodes):
+        self.graph = {node: [] for node in nodes}
+        self.reverse_graph = {node: [] for node in nodes}
+
+    def add_edge(self, edge_from, edge_to):
+        self.graph[edge_from].append(edge_to)
+        self.reverse_graph[edge_to].append(edge_from)
+
+    def get_follows(self, node):
+        return self.graph.get(node, [])
+
+    def get_dependencies(self, node):
+        return self.reverse_graph.get(node, [])
+
+    def topological_sort(self):
+        """ List DAG nodes in topological order. """
+
+        in_degree = {node: 0 for node in self.graph.keys()}
+
+        for i in self.graph:
+            for j in self.graph[i]:
+                in_degree[j] += 1
+
+        queue = []
```

Comment:
nit: I think `deque` should be favored in this case, if we consider time complexity.

Reply:
Got it